### PR TITLE
Add more test coverage

### PR DESCRIPTION
**Pull request by VirgileHlav**
Added more tests for raise/log cases and few untested functions in:
- `torchtext.nn.modules.multiheadattention.py`
- `torchtext.models.roberta.bundler.py`
- `torchtext.functional.py`

Following stats in [codecov.io/gh/pytorch/text](https://app.codecov.io/gh/pytorch/text).

### CODE REVIEW

File: test/test_functional.py [hunk 1]
```diff
@@ -68,6 +77,8 @@ def _add_token(self, test_scripting):
         func = functional.add_token
         if test_scripting:
             func = torch.jit.script(func)
+
+        # List[List[int]]
```
Comment (Nayef211) [inline test/test_functional.py:81]: nit: can we prefix these comments with `case:`. At a first glance it almost looks like commented code.

File: test/test_functional.py [hunk 2]
```diff
@@ -1,90 +1,78 @@
...
+        inputs, padding_value, expected_list = configs
         func = functional.to_tensor
         if test_scripting:
             func = torch.jit.script(func)
-        actual = func(input, padding_value=padding_value)
-        expected = torch.tensor([[1, 2, 0], [1, 2, 3]], dtype=torch.long)
-        torch.testing.assert_close(actual, expected)
+        else:
+            with self.assertRaises(TypeError):
+                func("test")
```
Comment (Nayef211) [inline test/test_functional.py:25]: Just wondering if this additional assertion helps to increase test coverage by ensuring the exception handling implementation is working correctly?

**Review by Nayef211 — APPROVED** [review]
Overall LGTM. Thanks for helping improve the test coverage for our repo!

File: test/test_functional.py (hunk 2 above)
Comment (VirgileHlav) [reply test/test_functional.py:25]: Yes, this is to test that an input type that is not a `List[int]` or `List[List[int]]` raises a `TypeError` (see [here](https://github.com/pytorch/text/blob/main/torchtext/functional.py#L37)).

File: test/data/test_modules.py [hunk 3]
```diff
@@ -213,11 +190,94 @@ def test_broadcast_scaled_dot_product(self):
...
-        # attn_mask's dim -3 is not equal to neither batch size or 1
-        with self.assertRaisesRegex(RuntimeError, "The size of the attn_mask is not correct."):
-            SDP(
-                query.expand(tgt_len, bsz * nhead, embed_dim),
-                key.expand(src_len, bsz * nhead, embed_dim),
-                value.expand(src_len, bsz * nhead, embed_dim),
-                attn_mask=attn_mask_2D.expand(2, tgt_len, src_len),
-            )
+
+    def test_assert_raises(self):
```
Comment (mthrok) [inline test/data/test_modules.py:194]: This test name is generic and thus it is making it hard to understand what test is about.

Also can you add docstring so to tell the intended behavior of the test? Typically this should be "(Under some condition) this feature should produce this result"

File: test/data/test_modules.py [hunk 4]
```diff
@@ -213,11 +190,94 @@ def test_broadcast_scaled_dot_product(self):
...
+        embed_dim, nhead, tgt_len, src_len, bsz = 10, 5, 6, 10, 64
+        SDP = ScaledDotProduct()
+        query = torch.rand((tgt_len, 1, embed_dim))
+        key = value = torch.rand((src_len, 1, embed_dim))
+        attn_mask_2D = torch.randint(0, 2, (tgt_len, src_len))
+
+        configs = [
```
Comment (mthrok) [inline test/data/test_modules.py:201]: Looking at the following `for` loop, the configs here are independent each other. However in the current implementation, one of the configs fail to pass the test, then the rest of the configs are not tested.

Unit tests should be independent each other. So these configs are better extracted out of the test logic, using `parameterized.expand`.

File: test/test_functional.py (hunk 2 above)
Comment (mthrok) [reply test/test_functional.py:25]: Looking at the flow of the test logic, this assertion kind of showed us suddenly. 

Also that assertion does not require the parameterized input, so why not put it in a stand-alone test?

File: test/test_functional.py [hunk 5]
```diff
@@ -1,90 +1,78 @@
...
+        [True, False],
+        [
+            [[[1, 2], [1, 2, 3]], [[1, 2], [1, 2]]],
+            [[1, 2, 3], [1, 2]],
+            [[["a", "b"], ["a", "b", "c"]], [["a", "b"], ["a", "b"]]],
+            [["a", "b", "c"], ["a", "b"]],
+        ]
+    )
+    def test_truncate(self, test_scripting, configs):
+        """test truncation on both sequence and batch of sequence with both str and int types"""
```
Comment (mthrok) [inline test/test_functional.py:41]: This docstring does not explain what the actual testing is. Could you rephrase in a way that better explains what the passing criteria is? such as something should match with somethig

File: test/test_functional.py [hunk 6]
```diff
@@ -1,90 +1,78 @@
...
+        inputs, expected = configs
         max_seq_len = 2
         func = functional.truncate
         if test_scripting:
             func = torch.jit.script(func)
+        else:
+            with self.assertRaises(TypeError):
+                func("test", max_seq_len=max_seq_len)
```
Comment (mthrok) [inline test/test_functional.py:49]: Since this assertion has nothing to do with the parameterized part of the input values, it should be in a separate, stand-alone test.

File: test/test_functional.py (hunk 2 above)
Comment (VirgileHlav) [reply test/test_functional.py:25]: @mthrok thanks for the comments! I addressed all the issues in #1667.

File: test/data/test_modules.py (hunk 3 above)
Comment (VirgileHlav) [reply test/data/test_modules.py:194]: Done. See #1667.

File: test/data/test_modules.py (hunk 4 above)
Comment (VirgileHlav) [reply test/data/test_modules.py:201]: Done. See #1667.

File: test/models/test_models.py [hunk 7]
```diff
@@ -135,3 +136,45 @@ def _train(model):
 
         self.assertEqual(model.encoder.state_dict(), encoder_current_state_dict)
         self.assertNotEqual(model.head.state_dict(), head_current_state_dict)
+
+    @pytest.fixture(autouse=True)
```
[the remark this answers is not in the document]
Comment (VirgileHlav) [reply test/models/test_models.py:140]: It does not, I changed the structure to use mock+patching. See #1667.

File: test/test_functional.py (hunk 5 above)
Comment (VirgileHlav) [reply test/test_functional.py:41]: Done. See #1667.

File: test/test_functional.py (hunk 6 above)
Comment (VirgileHlav) [reply test/test_functional.py:49]: Done. See #1667.
